### DISCUSSION TITLE
Struggling to decode bounding boxes

### DISCUSSION THREAD

**Discussion by martijnlammers** [ultralytics/ultralytics#9735] [Q&A] [answered]
Hi! I'm currently working on a side project using a yolov8 model from an onnx file to perform detections in C++. 

According to the documentation for yolov8, a feature vector consists of [x,y,w,h, prob1, prob2, prob3...] for each detection with the dimensions `batch size * bounding box + classes * possible detections` (1x8x8400 in my case).

When I try to decode the bounding boxes and probability from an inference result, I observe interesting things;
* A lot of probabilities exceed 100%
* There are detections where the width ~= height ~= x ~= y ~= probabilities. (line 2089)

[inference_result_log.txt](https://github.com/ultralytics/ultralytics/files/14885024/inference_result_log.txt)

So my questions would be; is this a normal/valid result, and if so, what steps would I require to do to get the normal bounding boxes/classes/probabilities from this inference?

**Comment by pderrenger** [comment]
Hello! It's great to hear about your project using YOLOv8 with C++. 😊

Based on what you’re experiencing, it looks like there might be an issue with the decoding or pre-processing of the model's output. YOLO's probabilities should indeed fall within the 0 to 1 range (or 0% to 100% if you prefer). Values exceeding 100% or having bounding box parameters (x, y, width, height) closely resembling each other or the probability values are certainly unusual.

A common step needed after obtaining the raw output from the model is applying a sigmoid function to the bounding box coordinates and objectness scores (and softmax to the class probabilities if not using auto-activation). This is because YOLO outputs are often logits or raw scores.

Here’s a quick rundown that might help:
1. **Apply transform**: Ensure that you apply the appropriate transformations (like sigmoid) to convert the raw scores to probabilities.
2. **Filter results**: Use a threshold to filter out detections with low probabilities (e.g., scores below 0.25 or another threshold that suits your scenario).

For decoding, the pseudo-code might look something like this (not actual C++ syntax):

```cpp
float sigmoid(float x) {
    return 1.0 / (1 + exp(-x));
}

for (auto &detection : detections) {
    float x = sigmoid(detection.x);
    float y = sigmoid(detection.y);
    float w = exp(detection.w); // assuming these are encoded as ln(width), ln(height)
    float h = exp(detection.h);
    // Transform probabilities similarly if they're logits
}

// Filter based on scores > threshold
```

🔗 If the issue persists, feel free to share more detailed inference logs or specific snippets of your decoding logic; perhaps there’s a minor discrepancy we can help identify.

Remember, the step to normalize (apply sigmoid) and interpret the raw outputs is crucial for YOLO models. I hope this points you in the right direction! Let us know how it goes.

**Comment by martijnlammers** [comment]
Thanks for the reply, really appreciate the help since I've been tinkering around with this for several days long.

I attempted to apply sigmoid to center_x, center_y and the class probabilities as you hinted, and apply the exp function on the width and height.

```C++
    std::vector<float> features; // originally 5.257160 15.202097 21.382442 25.222900 33.532639 51.217014 56.067261 58.087189....
    struct BoxParams
    {
        float center_x;
        float center_y;
        float width;
        float height;
    };

    struct Scores 
    {
        std::vector<float> values; // holds all the probabilities.
    };

    int rows = 8400;
    int cols = 8;

    std::vector<BoxParams> box_params_vec;
    std::vector<Scores> scores_vec;

    int pointer = 0;
    for(int i = 0; i < rows; i++)
    {
        pointer = i * cols;

        BoxParams params;
        params.center_x = Sigmoid(features[pointer + 0]);
        params.center_y = Sigmoid(features[pointer + 1]);
        params.width = exp(features[pointer + 2]);
        params.height = exp(features[pointer + 3]);

        Scores scores;
        for(int j = 4; j < cols; j++)
        {
            scores.values.push_back(Sigmoid(features[pointer + j]));
        }
        box_params_vec.push_back(params);
        scores_vec.push_back(scores);
    }
```
This resulted in the following logs: 
[result_after_sigmoid.txt](https://github.com/ultralytics/ultralytics/files/14890618/result_after_sigmoid.txt).

The preprocessing before feeding it to yolov8 consists of: 
* Standardizing 640x640 image pixels (greyscale)
* Zero padding to fit model (623x640 -> 640x640)
* Converting greyscale to RGB 

Any thoughts on what I could be doing wrong here?

**Reply by pderrenger** [reply]
@martijnlammers hey, thanks a lot for sharing your progress and the efforts you've put in so far! It sounds like you're on the right track with the sigmoid and exp transformations for the values. Based on what you've described and shared in your logs, your approach to adjusting center_x, center_y, width, height, and using sigmoid for class probabilities looks correct. 👍

A couple of things to check or consider that might help fine-tune the results further:

1. **Post-processing for Coordinates:** After applying sigmoid and exp, ensure that the bounding box coordinates (center_x, center_y) are scaled correctly according to the image size. Typically, you'd multiply `center_x` and `center_y` by the image width and height, respectively, to get the absolute positions in the image.

2. **Width and Height Adjustment:** Similar to coordinates, you would usually multiply the output of `exp(width)` and `exp(height)` by an anchor box dimension if your model uses anchor boxes. If not using anchor boxes, consider how the width and height relate to the entire image dimension.

3. **Thresholding Class Probabilities:** After applying sigmoid to class probabilities, it's common to filter out classes with probabilities below a certain threshold (e.g., 0.5) to reduce false positives.

4. **Greyscale to RGB Conversion:** Double-check how you're converting greyscale to RGB. Since each channel in a greyscale image is the same, duplicating the single greyscale channel to create three identical RGB channels should work fine. Ensure there’s no normalization issue (e.g., pixel value range being altered inadvertently during conversion).

5. **Normalization Check:** You mentioned standardizing pixel values. Make sure the normalization (e.g., dividing by 255 to get a [0, 1] range) matches what the model expects based on its training.

If after checking these aspects the issues persist, it could be worthwhile to visualize the bounding boxes on the image directly to see if there's an apparent pattern or anomaly in how they are being scaled or interpreted.

Keep up the great work, and feel free to reach out if you have more questions or run into further challenges!

**Comment by martijnlammers** [comment, marked as answer]
Thanks for the response. 

I found my problem, it was the pre-processing. When I was initially normalized the pixels, I coded it as `normalized_pixels.push_back(float(pixel / 255));`, and after careful inspection, I noticed the values were always zero.

After casting it properly as: `normalized_pixels.push_back(float(pixel) / 255.f);`, I managed to get expected results back.

**Reply by martijnlammers** [reply]
This also excluded the use of sigmoid/exp function.

**Reply by glenn-jocher** [reply]
Great catch on the pre-processing detail! 🎉 It's indeed crucial to cast the pixel values correctly to float before division to avoid integer truncation. This ensures the normalized pixel values fall within the expected [0.0, 1.0] range, prepping the image properly for inference.

Regarding your mention of excluding the sigmoid/exp function from the bounding box transform process, such transformations are typically applied to the network's raw outputs to map them to meaningful coordinate values and confidences. If you're handling raw predictions directly from the model without these adjustments, it might be necessary to apply them post-inference to interpret the results correctly.

For bounding box coordinates and class probabilities, they are often passed through a sigmoid function to bring them to a [0, 1] scale. The width and height (w, h) sometimes undergo an `exp` transform to map them from log space to actual size. This process varies depending on the model architecture and output format.

If you're working directly with raw ONNX outputs, consider reviewing the post-processing steps specific to YOLOv8 to ensure accurate interpretation of the bounding box and probability predictions. 🧐
